### PR TITLE
First prototype to show tasks on calendar and redirect to tasks app

### PR DESCRIPTION
**Pull request by bfritscher**
This is a start to address issue #28 . Feedback is welcome.

image

### CODE REVIEW

**Comment by bfritscher** [comment]
Currently changing the settings to display the tasks requires a refresh of the page. As the eventobjects of the current time ranges are already cached. Is there an easy way to force requery of the time ranges?

**Comment by georgehrke** [comment]
> 

There is a [fetchedTimeRanges store](https://github.com/nextcloud/calendar/blob/master/src/store/fetchedTimeRanges.js). You will have to add a mutation to reset the entire store. After that, if you increment the [modificationCount](https://github.com/nextcloud/calendar/blob/master/src/store/calendarObjects.js#L89), it will force fullCalendar to rerender and since there are no time-ranges, it will fetch them from the server.

**Comment by georgehrke** [comment]
@bfritscher Thank you very much for this pull-request :) 
I will give this a proper in-depth review later today.

File: src/models/calendarObject.js [hunk 1]
```diff
@@ -151,9 +151,29 @@ export default class CalendarObject {
 	getAllObjectsInTimeRange(start, end) {
 		const iterator = this.vcalendar.getVObjectIterator()
 		const firstVObject = iterator.next().value
-		if (!firstVObject) {
+		if (!firstVObject || (firstVObject.name === 'VTODO' && !firstVObject.dueTime)) {
 			return []
 		}
+		if (firstVObject.name === 'VTODO') {
+			// calendar-js is missing isAllDay, canModifyAllDay for TodoComponents
+			firstVObject.isAllDay = function() {
```
Comment (georgehrke) [inline src/models/calendarObject.js]: This is an ok workaround for now, but ideally we would fix this in [calendar-js](https://github.com/georgehrke/calendar-js). Let me take care of that: https://github.com/georgehrke/calendar-js/issues/73

File: css/fullcalendar.scss [hunk 2]
```diff
@@ -20,6 +20,10 @@
  *
  */
 
+.fc-event-nc-task .fc-content:before {
+	content: '✓';
```
[the remark this answers is not in the document]
Comment (jancborchardt) [reply css/fullcalendar.scss]: Fully agree, @georgehrke :)

**Comment by jancborchardt** [comment]
Only a superficial thing: It should be "Show tasks in calendar" (on → in) and it should be below "Enable birthday calendar" as it is sort of related since it adds a virtual kind of calendar.

**Comment by bfritscher** [comment]
Thanks for the feedback! 
I agree that an icon would be good to indentify it as a task entry, instead of just a checkmark. Not sure that it should be a checkbox which will always be uncheckd except if complete. Complete is already shown by the fact that it has strikethrough and less opacity. Google also displays the same icon: 
image 
image

For the naming and position of the settings entry I am open to suggestions, enable... or show... It is worth mentioning tos choose the phrasing that it is not only an additional virtual calendar. There are calendars with only events or only tasks, but there can also be calendars with events and calendars on the same calendar. In which case it could happend that no additional calendar is shown, and only new task entry appear on this calendar.

Thanks for the pointer on the timerange store.

**Comment by jancborchardt** [comment]
Hmm ok, then maybe test with just the `icon-checkmark` (as per https://docs.nextcloud.com/server/17/developer_manual/design/icons.html) just so it looks like Nextcloud-style instead of Unicode. :)

**Comment by bpcurse** [comment]
aCalendar (10Mio.+ downloads on gplay) uses unchecked checkboxes.
The checkmark might somehow be misleading to think that the task is done.

**Comment by jancborchardt** [comment]
@bpcurse could you show a screenshot?

**Comment by bpcurse** [comment]
@jancborchardt Sure, it looks like this (in dark mode):

Screenshot_aCalendar

Overdue items have a solid color (filled) checkbox with an exclamation mark in it.

**Comment by armaccloud** [comment]
Since tasks can have both a start- and a due date, it would be great if these task will actually occupy the space in the week view just like a calendar item does. This way you can visually see where in the day the task fits.

**Comment by jancborchardt** [comment]
I’d agree with @bpcurse that using non-filled checkboxes or filled ones depending on state is best – just like the Tasks app indeed does, so we have symmetry there.

@armaccloud this seems like something for a follow-up pull request.

**Comment by PyLit** [comment]
I am looking forward to this feature. I hope it gets figured out soon. I saw the new separation of calendars with and without task lists. I assume this is related and it got me excited.

**Comment by georgehrke** [comment]
Hey @bfritscher, 
Could you give us a quick update on the status of this PR please? :)

**Comment by georgehrke** [comment]
@bfritscher I would really like to include this in the next minor release. 
Would you mind if i take over? :)

**Comment by bfritscher** [comment]
@georgehrke I can set aside some time this evening to make an updated pull request.

Summary of the changes discussed above:
- page refresh reset fetchedTimeRange store and increment modificaitonCount
- icon show real state with a checkbox
- "Show tasks in calendar"  (did no get any other suggestion)
- if available use start/end date
- keep the current workaround for missing properties
- resolve merge conflicts

**Comment by bfritscher** [comment]
image

here a screenshot of second version with discussed changes

File: src/components/AppNavigation/Settings.vue [hunk 3]
```diff
@@ -30,6 +30,13 @@
 				@update:checked="toggleBirthdayEnabled">
 				{{ $t('calendar', 'Enable birthday calendar') }}
 			</ActionCheckbox>
+			<ActionCheckbox
+				class="settings-fieldset-interior-item"
+				:checked="showTasks"
+				:disabled="savingTasks"
+				@update:checked="toggleTasksEnabled">
+				{{ $t('calendar', 'Enable tasks in calendar') }}
```
Comment (juliushaertl) [inline src/components/AppNavigation/Settings.vue:38]:
```suggestion
				{{ $t('calendar', 'Show tasks in calendar') }}
```

**Comment by georgehrke** [comment]
Yes, it looked good when i checked it earlier.
I'm planning to have a `2.0.3` early next week and afterwards i will merge it.

**Comment by Eye-Of-Agamotto** [comment]
Hey guys I am relatively new to github. Is it possible for normal users to pull requests such as this to our own nextcloud instances. I am very excited for this feature

**Comment by GideonWolfe** [comment]
@Eye-Of-Agamotto I'm also quite excited, I've been waiting on this one since the day I installed NC. If you are running NC out of the git repo (unlikely) then you could check out this branch and test it.

Otherwise you'll need to wait until this merge gets packaged into whatever package you're using, (docker image, snap, etc)

**Comment by georgehrke** [comment]
> If you are running NC out of the git repo (unlikely) then you could check out this branch and test it.

Please do not clone git into your productive instance ...
Testing files and our javascript development dependencies will open up huge attack surfaces to compromise your instance / server.

File: templates/main.php [hunk 4]
```diff
@@ -21,3 +21,5 @@
  */
 script('calendar', 'calendar');
 style('calendar', 'calendar');
+
+?>
```
Comment (georgehrke) [inline templates/main.php:25]: Todo: remove me.

File: src/models/calendarObject.js [hunk 5]
```diff
@@ -151,9 +151,30 @@ export default class CalendarObject {
...
+				return true
+			}
+			if (!firstVObject.startDate) {
+				firstVObject.updatePropertyWithValue('DTSTART', DateTimeValue.fromJSDate(new Date(), true))
+				firstVObject.startDate = firstVObject.dueTime
+			}
+			firstVObject.endDate = firstVObject.dueTime
+
+			if (firstVObject.percent && !firstVObject._titleUpdated) {
+				firstVObject.title += ` (${firstVObject.percent}%)`
```
Comment (georgehrke) [inline src/models/calendarObject.js:172]: todo: this should me moved to the code that's creating the fc-events.

**Comment by georgehrke** [comment]
As discussed further above, I improved the support for VTODO in calendar-js, so we no longer need the hack in `/src/models/calendarObjects.js`: https://github.com/georgehrke/calendar-js/pull/131

I will take care of rebasing and resolving the small leftover ToDos tomorrow.

File: src/fullcalendar/eventClick.js [hunk 6]
```diff
@@ -33,27 +36,69 @@ import { getPrefixedRoute } from '../utils/router'
...
+/**
+ * Handle eventClick for VTODO
+ *
+ * @param {EventDef} event FullCalendar event
+ * @param {Object} store The Vuex store
+ * @param {Window} window The window object
+ */
+function handleToDoClick(event, store, window) {
+	if (!store.state.settings.tasksEnabled) {
+		showInfo(t('calendar', 'Please ask your administrator to enable the Tasks App.'))
```
Comment (tcitworld) [inline src/fullcalendar/eventClick.js:94]: Not applicable when public link.

**Comment by georgehrke** [comment]
@tcitworld Is this good to go from your side? :)

**Comment by georgehrke** [comment]
@bfritscher Thanks a lot for this contribution!

**Comment by bfritscher** [comment]
@georgehrke Thank you for the improvements and the merge!

**Comment by szaimen** [comment]
When will the next calendar release happen? Would love to test this great new feature :)

**Comment by georgehrke** [comment]
2.1.0 will probably come end of June / early July. I had too many other things on my plate last month and didn't managed to work as much on the Calendar as i hoped to.

**Pull request merged**